pandas/core/ops.py

```diff
@@ -821,54 +816,72 @@ def na_op(x, y):
 
         return result
 
-    def wrapper(self, other, axis=None):
-        # Validate the axis parameter
-        if axis is not None:
-            self._get_axis_number(axis)
-
-        if isinstance(other, ABCSeries):
-            name = _maybe_match_name(self, other)
-            if not self._indexed_same(other):
-                msg = 'Can only compare identically-labeled Series objects'
-                raise ValueError(msg)
-            return self._constructor(na_op(self.values, other.values),
-                                     index=self.index, name=name)
-        elif isinstance(other, ABCDataFrame):  # pragma: no cover
+    def wrapper(self, other):
+        if isinstance(other, ABCDataFrame):  # pragma: no cover
             return NotImplemented
-        elif isinstance(other, (np.ndarray, pd.Index)):
-            # do not check length of zerodim array
-            # as it will broadcast
-            if (not is_scalar(lib.item_from_zerodim(other)) and
-                    len(self) != len(other)):
-                raise ValueError('Lengths must match to compare')
 
-            if isinstance(other, ABCPeriodIndex):
-                # temp workaround until fixing GH 13637
-                # tested in test_nat_comparisons
-                # (pandas.tests.series.test_operators.TestSeriesOperators)
-                return self._constructor(na_op(self.values,
-                                               other.astype(object).values),
-                                         index=self.index)
+        elif isinstance(other, ABCSeries) and not self._indexed_same(other):
+            msg = 'Can only compare identically-labeled Series objects'
+            raise ValueError(msg)
 
-            return self._constructor(na_op(self.values, np.asarray(other)),
-                                     index=self.index).__finalize__(self)
+        res_name = _get_series_op_result_name(self, other)
 
-        elif isinstance(other, pd.Categorical):
-            if not is_categorical_dtype(self):
-                msg = ("Cannot compare a Categorical for op {op} with Series "
-                       "of dtype {typ}.\nIf you want to compare values, use "
-                       "'series <op> np.asarray(other)'.")
-                raise TypeError(msg.format(op=op, typ=self.dtype))
+        if is_timedelta64_dtype(self):
+            res = op(pd.TimedeltaIndex(self), other)
+            return self._constructor(res, index=self.index, name=res_name)
+
+        elif is_datetime64_dtype(self) or is_datetime64tz_dtype(self):
+            # kludge; DatetimeIndex refuses to compare against None or
+            # datetime.date; until the "right" behavior is resolved, we cast
+            # these types here to types that DatetimeIndex understand.
+            if type(other) is datetime.date:
```

Comment:
use isinstance.

what do you mean refuses to compare?

Reply:
> use isinstance.

Will change, but note the `isinstance` version is pretty verbose: `if isinstance(other, datetime.date) and not isinstance(other, datetime.datetime)`.

> what do you mean refuses to compare?

DatetimeIndex comparison with a) datetime.date or b) None raises TypeError; Series casts to datetime and NaT, respectively.

```
dti = pd.date_range('2016-01-01', periods=2)
ser = pd.Series(dti)

>>> dti ==  datetime.date(2016, 1, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/core/indexes/datetimes.py", line 119, in wrapper
    other = _ensure_datetime64(other)
  File "pandas/core/indexes/datetimes.py", line 145, in _ensure_datetime64
    raise TypeError('%s type object %s' % (type(other), str(other)))
TypeError: <type 'datetime.date'> type object 2016-01-01

>>> ser == datetime.date(2016, 1, 1)
0     True
1    False
dtype: bool

>>> dti != None
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/core/indexes/datetimes.py", line 119, in wrapper
    other = _ensure_datetime64(other)
  File "pandas/core/indexes/datetimes.py", line 145, in _ensure_datetime64
    raise TypeError('%s type object %s' % (type(other), str(other)))
TypeError: <type 'NoneType'> type object None

>>> ser != None
0    True
1    True
dtype: bool
```

So really one of these two behaviors should be chosen as canonical and the other changed (and tested), at which point these shims will be unnecessary.

Reply:
these comparison should work for both DTI and None. These work for scalars as well already.

Reply:
I am totally on board with the "DTI comparison should behave like a vectorized Timestamp comparison" approach, but it isn't 100% straightforward:

For `datetime.date` comparison, `Timestamp.__eq__` is always `False` and `Timestamp.__ne__` is always `True`; lt, le, gt, ge all raise `TypeError`.
```
dti = pd.date_range('2016-01-01', periods=2)
ts = dti[0]
dt = ts.to_pydatetime()
d = dt.date()

>>> ts == d
False
>>> ts < d
  File "pandas/_libs/tslibs/timestamps.pyx", line 121, in pandas._libs.tslibs.timestamps._Timestamp.__richcmp__
    raise TypeError('Cannot compare type %r with type %r' %
TypeError: Cannot compare type 'Timestamp' with type 'date'
```

Same behavior for `None`, i.e. it is _not_ treated being equivalent to `pd.NaT`.

Reply:
hmm, I would prefer a patch independetly of this one to fix these 2 cases, these already work with DTI

Reply:
#19301 is intended to do that.  Did you have something else in mind?

Reply:
yes saw that after